cds-snc/platform-forms-client
# Error boundary + error screens

# Summary | Résumé

- Exports Layout from the Base Layout so we can use it to wrap the Error Boundary
- Add  Error boundaries 
1) Root level - outside the `AccessControlProvider` - this uses the base layout wrapper
2) Page level - wraps the `<Component {...pageProps} />` this will be wrapped with the page Layout

- Update error pages for 404 ,Unauthorized + + 500 and other errors 
 
*Root level Error boundary*
<img width="500" alt="Screenshot 2023-05-11 at 1 09 47 PM" src="https://github.com/cds-snc/platform-forms-client/assets/62242/65a896bc-141f-4d87-8086-857f05ce88e5">

*In page / component level Error boundary*
<img width="700" alt="Screenshot 2023-05-11 at 10 55 52 AM" src="https://github.com/cds-snc/platform-forms-client/assets/62242/cf6a2d12-70b5-4a56-94e3-561178bad089">

*404*
<img width="700" alt="Screenshot 2023-05-11 at 1 18 05 PM" src="https://github.com/cds-snc/platform-forms-client/assets/62242/3bb908e5-ccfa-4c94-81b4-fca90ec7980e">

*Unauthorized*

<img width="700" alt="Screenshot 2023-05-11 at 1 21 25 PM" src="https://github.com/cds-snc/platform-forms-client/assets/62242/b7e6154d-2a60-4666-9105-5a98efc9e99b">

+ 500 and other errors are covered as well

Note from the old React docs there's quite a few cases where this won't catch.

Given it's hard to get the app to throw I created a debug component - see Test notes below.

https://legacy.reactjs.org/docs/error-boundaries.html#:~:text=Error%20boundaries%20are%20React%20components,the%20whole%20tree%20below%20them.



<img width="965" alt="Screenshot 2023-05-10 at 4 05 23 PM" src="https://github.com/cds-snc/platform-forms-client/assets/62242/afc2a461-3c07-4ac5-82b4-222112c44d72">

# Test instructions | Instructions pour tester la modification

To test the Error Boundary

see : https://github.com/cds-snc/platform-forms-client/pull/2089#issuecomment-1544422386

To test the error boundary 

```javascript
// add the TriggerError component to a page or higher up the stack
// visit a url and add ?forceError=true to the url
import { TriggerError } from "@components/globals";
...

<TriggerError />

```


# Unresolved questions / Out of scope | Questions non résolues ou hors sujet

Are there any related issues or tangent features you consider out of scope for
this issue that could be addressed in the future? If so please create issues and provide
links in this section

# Pull Request Checklist

Please complete the following items in the checklist before you request a review:
